casper/src/main/resources/Registry.rho

```diff
@@ -188,28 +152,64 @@ in {
         contract TreeHashMapSetter(@map, @nybList, @n, @len, @newVal, @suffix, ret) = {
           // Look up the value of the node at [map, nybList.slice(0, n + 1)
           new valCh, restCh in {
-            nodeGet!((map, nybList.slice(0, n)), *valCh) |
-            for (@val <- valCh) {
-              if (n == len) {
-                // If we're at the end of the path, set the node to newVal.
-                nodeSetMap!((map, nybList), suffix, newVal, *ret)
-              } else {
-                // Otherwise make the rest of the path exist.
-                // Bit k set means node k exists.
-                if ((val/powers.nth(nybList.nth(n))) % 2 == 0) {
-                  // Node missing
-                  // Create node, set node to 0
-                  MakeNode!(0, (map, nybList.slice(0, n + 1))) |
-                  new ack in {
-                    // Atomically update current node to val | (1 << nybList.nth(n))
-                    nodeSetBit!((map, nybList.slice(0, n)), nybList.nth(n), *ack) |
-                    for (_ <- ack) {
+            match (map, nybList.slice(0, n)) {
+              node => {
+                for (@val <<- @[node, *storeToken]) {
```

Comment:
Nice. Now the lock is only acquired if there's a value present at the node. Cool.